IllegalArgumentException: 'isReleased' is already in use

<!--- Provide a general summary of the issue in the Title above -->

Evaluation of the ReleaseOnceChannelPool in two classloaders with shared netty will lead to the exception 
```java
Caused by: java.lang.IllegalArgumentException: 'isReleased' is already in use
	at io.netty.util.ConstantPool.createOrThrow(ConstantPool.java:113)
	at io.netty.util.ConstantPool.newInstance(ConstantPool.java:95)
	at io.netty.util.AttributeKey.newInstance(AttributeKey.java:55)
	at software.amazon.awssdk.http.nio.netty.internal.ReleaseOnceChannelPool.<clinit>(ReleaseOnceChannelPool.java:38)
```
and 
```java
Caused by: java.lang.NoClassDefFoundError: Could not initialize class software.amazon.awssdk.http.nio.netty.internal.ReleaseOnceChannelPool
	at software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient$1.newPool(NettyNioAsyncHttpClient.java:158)
	at software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient$1.newPool(NettyNioAsyncHttpClient.java:143)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(Unknown Source)
	at software.amazon.awssdk.http.nio.netty.internal.SdkChannelPoolMap.get(SdkChannelPoolMap.java:42)
```

## Describe the bug
<!--- A clear and concise description of what the bug is -->

We use https://github.com/pf4j/pf4j for the system with netty server serving grpc with some pluggable handlers which are sharing netty in a classpath, but have anything else within an isolated classpath. If two plugins trying to work with dynamodb, each with own copy of the sdk, second plugin will get the exception above, trying to init

https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ReleaseOnceChannelPool.java#L38

at the moment its a *2.13.33* version.

## Expected Behavior
<!--- Tell us what should happen -->
Each sdk have isolated init from the shared netty part. Netty is pretty big framework and embedding it into each plugin could be a great waste of memory. So would be nice to have something like classloader id on such an id.

## Current Behavior
<!--- Tell us what happens instead of the expected behavior -->

<!--- Include full errors, uncaught exceptions, stack traces, and relevant logs -->
<!--- To turn on SDK logging, follow instructions here: http://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/java-dg-logging.html -->
<!--- If service responses are relevant, please include wirelogs -->

## Steps to Reproduce
Create a global classloader, load netty, create a child classloader (loading first from child, then from parent), load sdk in there, call sdk. Create the second classloader same as first child, repeat the sdk call.
<!--- Provide a self-contained, concise snippet of code that can be used to reproduce the issue -->
<!--- For more complex issues provide a repo with the smallest sample that reproduces the bug -->
<!--- Avoid including business logic or unrelated code, it makes diagnosis more difficult -->
The project we're using is the https://github.com/bsideup/liiklus.

## Possible Solution
<!--- Not required, but suggest a fix/reason for the bug -->
Adding classloader id could help, so that attributes wouldn't care about others.

## Context

Almost the same issue was fixed in the grpc https://github.com/grpc/grpc-java/issues/6707

<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment where the bug was discovered -->
* AWS Java SDK version used: 2.13.33
* JDK version used: 11
* Operating System and version: nix

--- a/build-tools/src/main/resources/software/amazon/awssdk/checkstyle.xml
+++ b/build-tools/src/main/resources/software/amazon/awssdk/checkstyle.xml
@@ -365,6 +365,14 @@
             <property name="ignoreComments" value="true"/>
         </module>
 
+        <!-- Checks that we don't use AttributeKey.newInstance directly -->
+        <module name="Regexp">
+            <property name="format" value="AttributeKey\.newInstance"/>
+            <property name="illegalPattern" value="true"/>
+            <property name="message" value="Use NettyUtils.getOrCreateAttributeKey to safely declare AttributeKeys"/>
+            <property name="ignoreComments" value="true"/>
+        </module>
+
         <!-- Checks that we don't use plural enum names -->
         <module name="software.amazon.awssdk.buildtools.checkstyle.PluralEnumNames"/>
 

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ChannelAttributeKey.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ChannelAttributeKey.java
@@ -26,6 +26,7 @@ import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.http.Protocol;
 import software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChannelPool;
 import software.amazon.awssdk.http.nio.netty.internal.http2.PingTracker;
+import software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils;
 
 /**
  * Keys for attributes attached via {@link io.netty.channel.Channel#attr(AttributeKey)}.
@@ -36,70 +37,71 @@ public final class ChannelAttributeKey {
     /**
      * Future that when a protocol (http/1.1 or h2) has been selected.
      */
-    public static final AttributeKey<CompletableFuture<Protocol>> PROTOCOL_FUTURE = AttributeKey.newInstance(
+    public static final AttributeKey<CompletableFuture<Protocol>> PROTOCOL_FUTURE = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.protocolFuture");
 
     /**
      * Reference to {@link Http2MultiplexedChannelPool} which stores information about leased streams for a multiplexed
      * connection.
      */
-    public static final AttributeKey<Http2MultiplexedChannelPool> HTTP2_MULTIPLEXED_CHANNEL_POOL = AttributeKey.newInstance(
-        "aws.http.nio.netty.async.http2MultiplexedChannelPool");
+    public static final AttributeKey<Http2MultiplexedChannelPool> HTTP2_MULTIPLEXED_CHANNEL_POOL =
+        NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.http2MultiplexedChannelPool");
 
     public static final AttributeKey<PingTracker> PING_TRACKER =
-        AttributeKey.newInstance("aws.http.nio.netty.async.h2.pingTracker");
+        NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.h2.pingTracker");
 
     public static final AttributeKey<Http2Connection> HTTP2_CONNECTION =
-        AttributeKey.newInstance("aws.http.nio.netty.async.http2Connection");
+        NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.http2Connection");
 
     public static final AttributeKey<Integer> HTTP2_INITIAL_WINDOW_SIZE =
-        AttributeKey.newInstance("aws.http.nio.netty.async.http2InitialWindowSize");
+        NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.http2InitialWindowSize");
 
     /**
      * Value of the MAX_CONCURRENT_STREAMS from the server's SETTING frame.
      */
-    public static final AttributeKey<Long> MAX_CONCURRENT_STREAMS = AttributeKey.newInstance(
+    public static final AttributeKey<Long> MAX_CONCURRENT_STREAMS = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.maxConcurrentStreams");
 
     /**
      * {@link AttributeKey} to keep track of whether we should close the connection after this request
      * has completed.
      */
-    static final AttributeKey<Boolean> KEEP_ALIVE = AttributeKey.newInstance("aws.http.nio.netty.async.keepAlive");
+    static final AttributeKey<Boolean> KEEP_ALIVE = NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.keepAlive");
 
     /**
      * Attribute key for {@link RequestContext}.
      */
-    static final AttributeKey<RequestContext> REQUEST_CONTEXT_KEY = AttributeKey.newInstance(
+    static final AttributeKey<RequestContext> REQUEST_CONTEXT_KEY = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.requestContext");
 
-    static final AttributeKey<Subscriber<? super ByteBuffer>> SUBSCRIBER_KEY = AttributeKey.newInstance(
+    static final AttributeKey<Subscriber<? super ByteBuffer>> SUBSCRIBER_KEY = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.subscriber");
 
-    static final AttributeKey<Boolean> RESPONSE_COMPLETE_KEY = AttributeKey.newInstance(
+    static final AttributeKey<Boolean> RESPONSE_COMPLETE_KEY = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.responseComplete");
 
     /**
      * {@link AttributeKey} to keep track of whether we have received the {@link LastHttpContent}.
      */
-    static final AttributeKey<Boolean> LAST_HTTP_CONTENT_RECEIVED_KEY = AttributeKey.newInstance(
+    static final AttributeKey<Boolean> LAST_HTTP_CONTENT_RECEIVED_KEY = NettyUtils.getOrCreateAttributeKey(
         "aws.http.nio.netty.async.lastHttpContentReceived");
 
-    static final AttributeKey<CompletableFuture<Void>> EXECUTE_FUTURE_KEY = AttributeKey.newInstance(
+    static final AttributeKey<CompletableFuture<Void>> EXECUTE_FUTURE_KEY = NettyUtils.getOrCreateAttributeKey(
             "aws.http.nio.netty.async.executeFuture");
 
-    static final AttributeKey<Long> EXECUTION_ID_KEY = AttributeKey.newInstance(
+    static final AttributeKey<Long> EXECUTION_ID_KEY = NettyUtils.getOrCreateAttributeKey(
             "aws.http.nio.netty.async.executionId");
 
     /**
      * Whether the channel is still in use
      */
-    static final AttributeKey<Boolean> IN_USE = AttributeKey.newInstance("aws.http.nio.netty.async.inUse");
+    static final AttributeKey<Boolean> IN_USE = NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.inUse");
 
     /**
      * Whether the channel should be closed once it is released.
      */
-    static final AttributeKey<Boolean> CLOSE_ON_RELEASE = AttributeKey.newInstance("aws.http.nio.netty.async.closeOnRelease");
+    static final AttributeKey<Boolean> CLOSE_ON_RELEASE = NettyUtils.getOrCreateAttributeKey(
+            "aws.http.nio.netty.async.closeOnRelease");
 
     private ChannelAttributeKey() {
     }

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/Http1TunnelConnectionPool.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/Http1TunnelConnectionPool.java
@@ -29,6 +29,7 @@ import io.netty.util.concurrent.Promise;
 import java.net.URI;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils;
 import software.amazon.awssdk.utils.Logger;
 import software.amazon.awssdk.utils.StringUtils;
 
@@ -37,7 +38,7 @@ import software.amazon.awssdk.utils.StringUtils;
  */
 @SdkInternalApi
 public class Http1TunnelConnectionPool implements ChannelPool {
-    static final AttributeKey<Boolean> TUNNEL_ESTABLISHED_KEY = AttributeKey.newInstance(
+    static final AttributeKey<Boolean> TUNNEL_ESTABLISHED_KEY = NettyUtils.getOrCreateAttributeKey(
             "aws.http.nio.netty.async.Http1TunnelConnectionPool.tunnelEstablished");
 
     private static final Logger log = Logger.loggerFor(Http1TunnelConnectionPool.class);

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ReleaseOnceChannelPool.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ReleaseOnceChannelPool.java
@@ -26,6 +26,7 @@ import io.netty.util.concurrent.SucceededFuture;
 import java.util.concurrent.atomic.AtomicBoolean;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChannelPool;
+import software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils;
 
 /**
  * Wrapper around a {@link ChannelPool} to protect it from having the same channel released twice. This can
@@ -35,7 +36,8 @@ import software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChan
 @SdkInternalApi
 public class ReleaseOnceChannelPool implements ChannelPool {
 
-    private static final AttributeKey<AtomicBoolean> IS_RELEASED = AttributeKey.newInstance("isReleased");
+    private static final AttributeKey<AtomicBoolean> IS_RELEASED = NettyUtils.getOrCreateAttributeKey(
+            "software.amazon.awssdk.http.nio.netty.internal.http2.ReleaseOnceChannelPool.isReleased");
 
     private final ChannelPool delegate;
 

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java
@@ -50,6 +50,7 @@ import software.amazon.awssdk.annotations.SdkTestInternalApi;
 import software.amazon.awssdk.http.Protocol;
 import software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey;
 import software.amazon.awssdk.http.nio.netty.internal.utils.BetterFixedChannelPool;
+import software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils;
 import software.amazon.awssdk.utils.Logger;
 import software.amazon.awssdk.utils.Validate;
 
@@ -72,13 +73,13 @@ public class Http2MultiplexedChannelPool implements ChannelPool {
     /**
      * Reference to the {@link MultiplexedChannelRecord} on a channel.
      */
-    private static final AttributeKey<MultiplexedChannelRecord> MULTIPLEXED_CHANNEL = AttributeKey.newInstance(
-        "software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChannelPool.MULTIPLEXED_CHANNEL");
+    private static final AttributeKey<MultiplexedChannelRecord> MULTIPLEXED_CHANNEL = NettyUtils.getOrCreateAttributeKey(
+            "software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChannelPool.MULTIPLEXED_CHANNEL");
 
     /**
      * Whether a parent channel has been released yet. This guards against double-releasing to the delegate connection pool.
      */
-    private static final AttributeKey<Boolean> RELEASED = AttributeKey.newInstance(
+    private static final AttributeKey<Boolean> RELEASED = NettyUtils.getOrCreateAttributeKey(
         "software.amazon.awssdk.http.nio.netty.internal.http2.Http2MultiplexedChannelPool.RELEASED");
 
     private final ChannelPool connectionPool;

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java
@@ -16,6 +16,7 @@
 package software.amazon.awssdk.http.nio.netty.internal.utils;
 
 import io.netty.channel.EventLoop;
+import io.netty.util.AttributeKey;
 import io.netty.util.concurrent.EventExecutor;
 import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.GenericFutureListener;
@@ -159,5 +160,17 @@ public final class NettyUtils {
                                           + "AWS SDK for Java team on GitHub, because it could result in race conditions.");
             log.warn(() -> "Execution is happening outside of the expected event loop.", exception);
         }
+    }
+
+    /**
+     * @return an {@code AttributeKey} for {@code attr}. This returns an existing instance if it was previously created.
+     */
+    public static <T> AttributeKey<T> getOrCreateAttributeKey(String attr) {
+        if (AttributeKey.exists(attr)) {
+            return AttributeKey.valueOf(attr);
+        }
+        //CHECKSTYLE:OFF - This is the only place allowed to call AttributeKey.newInstance()
+        return AttributeKey.newInstance(attr);
+        //CHECKSTYLE:ON
     }
 }

--- a/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/OrderedWriteChannelHandlerContext.java
+++ b/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/OrderedWriteChannelHandlerContext.java
@@ -31,7 +31,7 @@ import software.amazon.awssdk.annotations.SdkInternalApi;
 @SdkInternalApi
 public class OrderedWriteChannelHandlerContext extends DelegatingChannelHandlerContext {
     private static final AttributeKey<Void> ORDERED =
-        AttributeKey.newInstance("aws.http.nio.netty.async.OrderedWriteChannelHandlerContext.ORDERED");
+        NettyUtils.getOrCreateAttributeKey("aws.http.nio.netty.async.OrderedWriteChannelHandlerContext.ORDERED");
 
     private OrderedWriteChannelHandlerContext(ChannelHandlerContext delegate) {
         super(delegate);

--- a/http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtilsTest.java
+++ b/http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtilsTest.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal.utils;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.netty.util.AttributeKey;
+import org.junit.Test;
+
+public class NettyUtilsTest {
+    @Test
+    public void testGetOrCreateAttributeKey_calledTwiceWithSameName_returnsSameInstance() {
+        String attr = "NettyUtilsTest.Foo";
+        AttributeKey<String> fooAttr = NettyUtils.getOrCreateAttributeKey(attr);
+        assertThat(NettyUtils.getOrCreateAttributeKey(attr)).isSameAs(fooAttr);
+    }
+}

